project/Component/NefSwiftPlayground/SwiftPlayground.swift

```diff
@@ -0,0 +1,170 @@
+            yield: repos.get)^.reportStatus(step: step, in: env.shell.out, verbose: true)
+        }
+    }
+    
+    private func modules(step: Step, repos: [String], excludes: [PlaygroundExcludeItem]) -> EnvIO<Shell, SwiftPlaygroundError, [Module]> {
+        EnvIO { (out, shell) in
+            let modules = IOPartial<SwiftPlaygroundError>.var([Module].self)
+            
+            return binding(
+                       |<-out.printStep(step: step, information: "Get modules from repositories..."),
+               modules <- self.swiftLibraryModules(in: repos, excludes: excludes).provide(shell),
+            yield: modules.get)^.reportStatus(step: step, in: out, verbose: true)
+        }
+    }
+    
+    private func swiftPlayground(step: Step, modules: [Module], path: PlaygroundResolutionPath) -> EnvIO<PlaygroundEnvironment, SwiftPlaygroundError, Void> {
+        EnvIO { env in
+            binding(
+                |<-env.shell.out.printStep(step: step, information: "Building Swift Playground..."),
+                |<-self.buildPlaygroundBook(modules: modules, playgroundPath: path.playgroundPath).provide(env.system),
+            yield: ())^.reportStatus(step: step, in: env.shell.out, verbose: false)
+        }
+    }
+    
+    // MARK: steps <helpers>
+    private func removeItem(at itemPath: String, useCache: Bool = false) -> EnvIO<FileSystem, SwiftPlaygroundError, Void> {
+        EnvIO { system in
+            guard !useCache else { return IO.pure(()) }
+            
+            let removeFileIO = system.remove(itemPath: itemPath).mapLeft { _ in SwiftPlaygroundError.clean(item: itemPath) }^
+            return system.exist(itemPath: itemPath) ? removeFileIO : IO.pure(())
+        }
+    }
+    
+    private func makeStructure(buildPath: String) -> EnvIO<FileSystem, SwiftPlaygroundError, Void> {
+        EnvIO { system in
+            system.createDirectory(atPath: buildPath)^
+                  .mapLeft { _ in .structure }
+        }
+    }
+    
+    private func buildPackage(content: String, packageFilePath: String, packagePath: String, buildPath: String) -> EnvIO<(FileSystem, PlaygroundShell), SwiftPlaygroundError, Void> {
+        EnvIO { (system, shell) in
+            let writePackageIO = system.write(content: content, toFile: packageFilePath).mapLeft { _ in SwiftPlaygroundError.checkout }
+            let resolvePackageIO = shell.resolve(packagePath: packagePath, buildPath: buildPath).mapLeft { _ in SwiftPlaygroundError.checkout }
+
+            return writePackageIO.followedBy(resolvePackageIO)
+        }
+    }
+    
+    private func repositories(checkoutPath: String) -> EnvIO<FileSystem, SwiftPlaygroundError, [String]> {
+        EnvIO { fileSystem in
+            fileSystem.items(atPath: checkoutPath)
+                      .mapLeft { _ in .checkout }
+                      .map { repos in repos.filter { repo in !repo.filename.contains("swift-") } }
```

Comment:
This looks like we need to add a `bimap` method on `IO` and `EnvIO`.

Reply:
exactly it I thought yesterday during the schedules revision - `dimap`...